backend/mora/auth/sso.py

```diff
@@ -5,38 +5,176 @@
+from onelogin.saml2.utils import OneLogin_Saml2_Utils
 
 from . import base
-from . import tokens
-from .. import settings
 
 basedir = os.path.dirname(__file__)
 
+blueprint = flask.Blueprint('sso', __name__, static_url_path='',
+                            url_prefix='/saml')
+
+
+def get_saml_settings(app):
+    config = app.config
+
+    insecure = config['SAML_IDP_INSECURE']
+    cert_file = config['SAML_CERT_FILE']
+    key_file = config['SAML_KEY_FILE']
+    requests_signed = config['SAML_REQUESTS_SIGNED']
+    saml_idp_metadata_url = config['SAML_IDP_METADATA_URL']
+
+    remote = OneLogin_Saml2_IdPMetadataParser.parse_remote(
+        saml_idp_metadata_url,
+        validate_cert=not insecure
+    )
+
+    s = {
+        "strict": True,
+        "debug": True,
+        "sp": {
+            "entityId": flask.url_for('sso.metadata', _external=True),
+            "assertionConsumerService": {
+                "url": flask.url_for('sso.acs', _external=True),
+                "binding": "urn:oasis:names:tc:SAML:2.0:bindings:HTTP-POST"
+            },
+            "singleLogoutService": {
+                "url": flask.url_for('sso.sls', _external=True),
+                "binding": "urn:oasis:names:tc:SAML:2.0:bindings:HTTP-Redirect"
+            },
+        },
+        "security": {
+            "authnRequestsSigned": requests_signed,
+            "logoutRequestSigned": requests_signed
+        }
+    }
+
+    s.setdefault('sp', {}).update(remote.get('sp'))
+    s.setdefault('idp', {}).update(remote.get('idp'))
+
+    if requests_signed:
+        with open(cert_file, 'r') as cf:
+            cert = cf.read()
+        with open(key_file, 'r') as kf:
+            key = kf.read()
+
+        s['sp'].update({
+            "x509cert": cert,
+            "privateKey": key
+        })
+
+    return s
+
+
+def prepare_flask_request():
+    # If server is behind proxys or balancers use the HTTP_X_FORWARDED fields
+    url_data = parse.urlparse(flask.request.url)
+    return {
+        'https': 'on' if flask.request.scheme == 'https' else 'off',
+        'http_host': flask.request.host,
+        'server_port': url_data.port,
+        'script_name': flask.request.path,
+        'get_data': flask.request.args.copy(),
+        'post_data': flask.request.form.copy()
+    }
+
+
+def init_saml_auth(func):
+    @functools.wraps(func)
+    def wrapper():
+        app = flask.current_app
+        config = app.extensions.get('saml')
+        if not config:
+            config = get_saml_settings(app)
+            app.extensions['saml'] = config
+        req = prepare_flask_request()
+        auth = OneLogin_Saml2_Auth(req, config)
+        return func(auth)
+    return wrapper
+
+
+@blueprint.route('/metadata/')
+@init_saml_auth
+def metadata(auth):
+    settings = auth.get_settings()
+    sp_metadata = settings.get_sp_metadata()
+    errors = settings.validate_metadata(sp_metadata)
+
+    if len(errors) == 0:
```

Comment:
As earlier, and I'd consider flipping the condition both locations — that is, have the check for errors be explicit:
```python
if errors:
    return flask.jsonify(errors), 401

resp = …
```